command/views/view.go

```diff
@@ -0,0 +1,122 @@
+// no-op configSources callback.
+func NewView(streams *terminal.Streams) *View {
+	return &View{
+		streams: streams,
+		colorize: &colorstring.Colorize{
+			Colors:  colorstring.DefaultColors,
+			Disable: true,
+			Reset:   true,
+		},
+		configSources: func() map[string][]byte { return nil },
+	}
+}
+
+// EnableColor controls the colorize implementation used by this view (and any
+// other views which depend on it). This is called by the code which processes
+// the global -no-color CLI flag, which happens after the view is initialized.
+func (v *View) EnableColor(color bool) {
+	v.colorize.Disable = !color
+}
+
+// SetConfigSources overrides the default no-op callback with a new function
+// pointer, and should be called when the config loader is initialized.
+func (v *View) SetConfigSources(cb func() map[string][]byte) {
+	v.configSources = cb
+}
+
+// Diagnostics renders a set of warnings and errors in human-readable form.
+// Warnings are printed to stdout, and errors to stderr.
+func (v *View) Diagnostics(diags tfdiags.Diagnostics) {
+	diags.Sort()
+
+	if len(diags) == 0 {
+		return
+	}
+
+	diags = diags.ConsolidateWarnings(1)
+
+	// Since warning messages are generally competing
+	if v.compactWarnings {
+		// If the user selected compact warnings and all of the diagnostics are
+		// warnings then we'll use a more compact representation of the warnings
+		// that only includes their summaries.
+		// We show full warnings if there are also errors, because a warning
+		// can sometimes serve as good context for a subsequent error.
+		useCompact := true
+		for _, diag := range diags {
+			if diag.Severity() != tfdiags.Warning {
+				useCompact = false
+				break
+			}
+		}
+		if useCompact {
+			msg := format.DiagnosticWarningsCompact(diags, v.colorize)
+			msg = "\n" + msg + "\nTo see the full warning notes, run Terraform without -compact-warnings.\n"
+			v.output(msg)
+			return
+		}
+	}
+
+	for _, diag := range diags {
+		var msg string
+		if v.colorize.Disable {
+			msg = format.DiagnosticPlain(diag, v.configSources(), v.streams.Stderr.Columns())
+		} else {
+			msg = format.Diagnostic(diag, v.configSources(), v.colorize, v.streams.Stderr.Columns())
+		}
+
+		if diag.Severity() == tfdiags.Error {
+			fmt.Fprint(v.streams.Stderr.File, msg)
+		} else {
+			fmt.Fprint(v.streams.Stdout.File, msg)
+		}
+	}
+}
+
+// HelpPrompt is intended to be called from commands which fail to parse all
+// of their CLI arguments successfully. It refers users to the full help output
+// rather than rendering it directly, which can be overwhelming and confusing.
+func (v *View) HelpPrompt(command string) {
+	fmt.Fprintf(v.streams.Stderr.File, helpPrompt, command)
+}
+
+const helpPrompt = `
+For more help on using this command, run:
+  terraform %s -help
+`
+
+// output is a shorthand for the common view operation of printing a string to
+// the stdout stream, followed by a newline.
+func (v *View) output(s string) {
+	fmt.Fprintln(v.streams.Stdout.File, s)
```

Comment:
Given how often we seemed to do `c.Ui.Output(fmt.Sprintf(...))` I wonder about making this have the signature `output(fmt string, args ...interface{})` instead, and then using `fmt.Fprintf` in here, cause I think in practice _that_ has been the more common case.

Building the formatting into the function might also help nudge towards the more correct interaction with `Colorize`, where we need to be colorizing the format string rather than the format result in order to avoid interpreting color-sequence-looking things in the format arguments:

```
view.output(v.colorize.Colorize("Something: [bold]%s[reset]\n"), something)
```

(As you recently saw, we've got real existing examples of us getting this wrong in the current setup where formatting is separated from output, where the colorize was applied to the formatting result rather than the format string itself.)

---

Separate thought but also related to this particular function:

I guess printing to the output streams is going to be a common enough operation in the views that it might justify having `fmt`-like helper functions directly on the `Streams` type for it:

```go
    v.streams.Printf("Something: %s\n", something) // wraps fmt.Fprintf(v.streams.Stdout.File, ...)
    v.streams.Println("Something ", something) // wraps fmt.Fprintln(v.streams.Stdout.File, ...)
    v.streams.EPrintf("Something: %s\n", something) // wraps fmt.Fprintf(v.streams.Stderr.File, ...)
    v.streams.EPrintln("Something ", something) // wraps fmt.Fprintln(v.streams.Stderr.File, ...)
```

I do want to be cautious about adding too much fluffy abstraction into the `terminal` package, but given that these are just thin convenience wrappers around an existing abstraction already in the standard library it feels like an okay compromise to me: I think we can assume that most folks who are familiar with Go will understand the analogy to `fmt.Printf`, `fmt.Println`, etc (and similar methods in `log`) and thus not have to think too hard to guess what these functions are going to do.

Reply:
I agree with these suggestions, and would like to add more helper functions to prevent the need for views to poke into the streams directly. One concern I have with changing the semantics of `output` to use a format string is that it is convenient to have a `Println` equivalent rather than scattering newlines around the call sites. This leads me towards the latter approach you suggested, although I'd like to leave that for a separate PR—this one is already too long.

Reply:
In the interests of connecting these dots together for potential future reference, we subsequently added these `fmt`-like functions over in #27759.